autometa/common/external/samtools.py

```diff
@@ -42,49 +41,63 @@ def sort(sam, bam, nproc=mp.cpu_count()):
 
     Raises
     ------
     TypeError
-        nproc must be an integer greater than zero
+        cpus must be an integer greater than zero
     FileNotFoundError
         Specified path is incorrect or the file is empty
-    ChildProcessError
-        Samtools did not run successfully, returns the return code from subprocessing call
+    subprocess.CalledProcessError
+        Samtools did not run successfully
     """
 
-    if type(nproc) is not int or nproc <= 0:
-        raise TypeError(f"nproc must be an integer greater than zero! Given: {nproc}")
-    if not os.path.exists(sam) or os.stat(sam).st_size == 0:
+    if type(cpus) is not int or cpus <= 0:
+        raise TypeError(f"cpus must be an integer greater than zero! Given: {cpus}")
+    if not os.path.exists(sam) or not os.path.getsize(sam):
         raise FileNotFoundError(
             f"The specified path: {sam} is either incorrect or the file is empty"
         )
-    samtools_out_dir = os.path.dirname(os.path.abspath(bam))
-    err = os.path.join(samtools_out_dir, "samtools.err")
-    out = os.path.join(samtools_out_dir, "samtools.out")
-    with tempfile.TemporaryDirectory() as tempdir:
-        temp_bam = os.path.join(tempdir, os.path.basename(bam))
-        cmd = (
-            f"samtools view -@{nproc} -bS {sam} | samtools sort -@{nproc} -o {temp_bam}"
+    with tempfile.TemporaryDirectory() as tempdir:  # this will delete the temporary files even if program stops in between
+        sort_fpath = os.path.join(tempdir, os.path.basename(bam))
+        # See https://stackoverflow.com/questions/13332268/how-to-use-subprocess-command-with-pipes
+        cmd_view = [
+            "samtools",
+            "view",
+            "-@",
+            str(cpus),
+            "-bS",
+            sam,
+        ]
+        cmd_sort = [
+            "samtools",
+            "sort",
+            "-@",
+            str(cpus),
+            "-o",
+            sort_fpath,
+        ]
+        logger.debug(" ".join(cmd_view))
+        logger.debug(" ".join(cmd_sort))
+        view = subprocess.Popen(
+            cmd_view, stdout=subprocess.PIPE, stderr=subprocess.PIPE
         )
-        logger.debug(f"cmd: {cmd}")
-        with open(out, "w") as stdout, open(err, "w") as stderr:
-            retcode = subprocess.call(cmd, stdout=stdout, stderr=stderr, shell=True)
-        if retcode != 0:
-            raise ChildProcessError(f"Sort failed, retcode: {retcode}")
-        else:
-            shutil.move(temp_bam, bam)
-        os.remove(err)
-        os.remove(out)
-
-
-def main(args):
-    sort(args.sam, args.bam, args.nproc)
+        # See https://stackoverflow.com/questions/34147353/python-subprocess-chaining-commands-with-subprocess-run
+        # Popen starts the process and carries on while run starts it and waits for it to finish.
+        subprocess.run(
+            cmd_sort,
+            stdin=view.stdout,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
```

Comment:
Should both of these be piped? Should stderr be directed to `devnull`? Ideally if `samtools` fails, it would be useful to have this error information available to the user so they know _why_ samtools failed.

Reply:
You are right both should not be piped. 
I think we have 2 options to convey the error to the user:
- Create a `samtools.err` file. However, in all the other scripts we have just directed the `stderr` to the `DEVNULL`. 
- Log the error so that the user can look it up if he wants to.
I prefer the latter one as it maintains consistency and would reduce the creation of new files.

Reply:
Even though I have implemented to log the error after `sort` , I am not sure if `check=True` would allow that. I think `check=True` would raise an error and the logging would never happen. I could not verify this as I was unable to explicitly cause `sort` to crash.